src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1792,20 +1829,112 @@ private void setItemToCacheFromRefresh(final RefreshItem refreshItem,
             logReturnedToken(request, result);
 
             // Update for cache key
+            // If token is refreshed with Family token, make sure the tokenCacheItem doesn't
+            // contain the client id. 
+            final TokenCacheItem tokenCacheItem = new TokenCacheItem(request, result, refreshItem.mMultiResource);
+            if (tokenRefreshedWithFRT(request, refreshItem)) {
+                tokenCacheItem.setClientId(null);
+            } 
+            
             mTokenCacheStore.setItem(refreshItem.mKey, new TokenCacheItem(request, result,
                     refreshItem.mMultiResource));
-
             setItemToCache(request, result, false);
         }
     }
 
-    private void removeItemFromCache(final RefreshItem refreshItem) {
+    /**
+     * Remove token from cache. 
+     */
+    private void removeItemFromCache(final AuthenticationRequest authenticationRequest, final RefreshItem refreshItem) {
+        // Token removal logic:
+        // 1) If refresh with resource specific token cache entry, clear RT with key(R,C,U,A)
```

Comment:
These comments should be moved over (or atleast copied over)  to the remove* methods below.